File: python/05.py
import re

filename = "05.txt"

strings = [x.strip() for x in open(filename).readlines()]

# Part 1
print(len([s for s in strings if (re.search(r'([aeiou].*){3,}', s) and
                                  re.search(r'(.)\1', s) and
                                  not re.search(r'ab|cd|pq|xy', s))]))

# Part 2
print(len([s for s in strings if (re.search(r'(.).\1', s) and
                                  re.search(r'(..).*\1', s)
                                 )]))

File: javascript/05.js
var fs = require('fs');
var regex = require('regex');
var filename = '05.txt';

var strings = fs.readFileSync(filename).toString().trim().split('\n');

// part 1
function hasThreeVowels(line) {
  return /([aeiou].*){3,}/ig.test(line);
}

function hasDoubles(line) {
  return /(.)\1/.test(line);
}

function isNotNaughty(line) {
  return !(/ab|cd|pq|xy/.test(line));
}

console.log(strings.reduce(function (total, line) {
  return total + ((hasThreeVowels(line) && hasDoubles(line) && isNotNaughty(line)) ? 1 : 0);
}, 0));



// part 2
function nonOverlappingPairs(line) {
  return /(.).\1/ig.test(line);
}

function specialRepeat(line) {
  return /(..).*\1/ig.test(line);
}

console.log(strings.reduce(function(total, line) {
  return total + ((nonOverlappingPairs(line) && specialRepeat(line)) ? 1 : 0);
}, 0));
